lib/msf/ui/console/command_dispatcher/core.rb

```diff
@@ -1794,6 +1836,177 @@ def cmd_sessions(*args)
+        return false
+      end
+      operator = checkin_searches[0].split(":")[1]
+      unless VALID_OPERATORS.include?(operator)
+        print_error("Please specify before or after for checkin query. Ex: last_checkin:before:1m30s. Given: #{operator}")
+        return false
+      end
+    end
+    if checkin_searches.length > 2
+      print_error("Too many checkin searches. Max: 2. Given: #{checkin_searches.length}")
+      return false
+    elsif checkin_searches.length == 2
+      field1, operator1, value1 = checkin_searches[0].split(":")
+      field2, operator2, value2 = checkin_searches[1].split(":")
+      unless VALID_OPERATORS.include?(operator1) && VALID_OPERATORS.include?(operator2)
+        print_error("last_checkin can only be searched for using before or after. Ex: last_checkin:before:1m30s")
+        return false
+      end
+      if operator1 == operator2
+        print_error("Cannot search for last_checkin with two #{operator1} arguments.")
+        return false
+      end
+      if (operator1 == 'before' && parse_duration(value2) < parse_duration(value1)) || (operator1 == 'after' && parse_duration(value1) < parse_duration(value2))
+        print_error("After value must be a larger duration than the before value.")
+        return false
+      end
+    end    
+    true
+  end
+
+  def filter_sessions_by_search(search_term)
+    matching_sessions = {}
+    field, operator, value = search_term.split(":")
+    framework.sessions.each do |session_id, session|
+      unless session.respond_to?(:last_checkin)
+        next if field == LAST_CHECKIN
+      end
+      matches_search = evaluate_search_criteria(session, search_term)
+      return nil if matches_search.nil?
+      case field
+      when LAST_CHECKIN
+        if session.last_checkin && evaluate_search_criteria(session, search_term)
+          matching_sessions[session_id] = session
+        end
+      when SESSION_TYPE, SESSION_ID
+        matching_sessions[session_id] = session if evaluate_search_criteria(session, search_term)
+      else
+        print_error("Unrecognized search term: #{field}")
+        return nil
+      end
+    end
+    matching_sessions
+  end
+
+  def evaluate_search_criteria(session, search_term)
+    field, operator, value = search_term.split(":")
+
+    case field
+    when LAST_CHECKIN
+      checkin_time = session.last_checkin
+      offset = parse_duration(value)
+      return nil unless offset
+      threshold_time = Time.now - parse_duration(value)
+      case operator
+      when "before"
+        return checkin_time < threshold_time
+      when "after"
+        return checkin_time > threshold_time
+      end
+    when SESSION_ID
+      return session.sid.to_s == operator
+    when SESSION_TYPE
+      return session.type.casecmp?(operator)
+    end
+  end
+
+  def parse_duration(duration)
+    total_time = 0
+    time_tokens = duration.scan(/(?:\d+\.?\d*|\.\d+)/).zip(duration.scan(/[a-zA-Z]+/))
```

Comment:
What happens here if the user puts in capital letters? I don't see them being downcase'd anywhere and the switch statement only deals with the lower case characters

Reply:
added a downcase to the switch 👍